File: OsmAnd/src/net/osmand/plus/widgets/TextViewEx.java
package net.osmand.plus.widgets;

import static net.osmand.plus.utils.FontCache.FONT_WEIGHT_NORMAL;

import android.content.Context;
import android.content.res.Resources.Theme;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import net.osmand.plus.R;
import net.osmand.plus.utils.FontCache;


public class TextViewEx extends androidx.appcompat.widget.AppCompatTextView {


	public TextViewEx(@NonNull Context context) {
		this(context, null);
	}

	public TextViewEx(@NonNull Context context, @Nullable AttributeSet attrs) {
		this(context, attrs, 0);

		parseAttributes(this, attrs, 0, 0);
	}

	public TextViewEx(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);

		parseAttributes(this, attrs, defStyleAttr, 0);
	}

	protected static void parseAttributes(@NonNull TextView textView, @Nullable AttributeSet attrs,
	                                      int defStyleAttr, int defStyleRes) {
		if (attrs != null) {
			Theme theme = textView.getContext().getTheme();
			TypedArray attributes = theme.obtainStyledAttributes(attrs, R.styleable.TextViewEx, defStyleAttr, defStyleRes);
			applyAttributes(textView, attributes);
			attributes.recycle();
		}
	}

	private static void applyAttributes(@NonNull TextView textView, @NonNull TypedArray attributes) {
		applyTypefaceWeight(textView, attributes);
	}

	private static void applyTypefaceWeight(@NonNull TextView textView, @NonNull TypedArray attributes) {
		if (!attributes.hasValue(R.styleable.TextViewEx_typefaceWeight) || textView.isInEditMode()) {
			return;
		}
		int weight = attributes.getInteger(R.styleable.TextViewEx_typefaceWeight, FONT_WEIGHT_NORMAL);

		Typeface typeface = textView.getTypeface();
		textView.setTypeface(FontCache.getFont(typeface, weight));
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/ui/views/TextViewEx.kt
package net.osmand.telegram.ui.views

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import net.osmand.telegram.R
import net.osmand.telegram.helpers.FontCache

class TextViewEx : AppCompatTextView {

	constructor(context: Context) : super(context)

	constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
		parseAttrs(attrs, 0)
	}

	constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
		context,
		attrs,
		defStyleAttr
	) {
		parseAttrs(attrs, defStyleAttr)
	}

	private fun parseAttrs(attrs: AttributeSet?, defStyleAttr: Int) {
		if (attrs == null) {
			return
		}

		val resolvedAttrs =
			context.theme.obtainStyledAttributes(attrs, R.styleable.TextViewEx, defStyleAttr, 0)
		applyAttrTypeface(resolvedAttrs)
		resolvedAttrs.recycle()
	}

	private fun applyAttrTypeface(resolvedAttrs: TypedArray) {
		if (isInEditMode || !resolvedAttrs.hasValue(R.styleable.TextViewEx_typeface)) {
			return
		}

		val typefaceName = resolvedAttrs.getString(R.styleable.TextViewEx_typeface)
		val typeface = FontCache.getFont(context, typefaceName!!)
		if (typeface != null) {
			val style = getTypeface()?.style ?: 0
			setTypeface(typeface, style)
		}
	}
}
